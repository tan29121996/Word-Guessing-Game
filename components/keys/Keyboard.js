import { View, Text, TouchableWithoutFeedback, ImageBackground } from "react-native";
import TouchableScale from 'react-native-touchable-scale';
import styles, { keyWidth } from "./Keyboard.styles";

const Keyboard = ({
  onKeyPressed = () => {},
}, {testID}) => {

  const keys = [
    ["w", "e", "r", "t", "y", "o", "q"],
    ["a", "s", "g", "u", "h", "i", "p"],
    ["z", "d", "f", "j", "l", "k", "x"],
    ["↵", "c", "v", "b", "n", "m", "←"],
  ];

  handleViewRef = ref => this.view = ref;
  
  bounce = () => this.view.bounce(800);

  return (
    <View style={styles.keyboard} testID='keyboard'>
      {keys.map((keyRow, i) => (
        <View style={styles.row} key={`row-${i}`}>
          {keyRow.map((key) => (
            <TouchableScale 
              onPress={() => onKeyPressed(key)} 
              key={key} 
              testID={key}
            >
              <ImageBackground
                resizeMode="cover"
                style={{
                  ...styles.letter,
                }}
                source={require('../img/option_letter.png')}>
                <View style={styles.characterContainer}>
                  <Text style={styles.character}>
                    {key.toUpperCase()}
                  </Text>
                </View>
              </ImageBackground>
            </TouchableScale>
          ))}
        </View>
      ))}
    </View>
  );
};

export default Keyboard;
