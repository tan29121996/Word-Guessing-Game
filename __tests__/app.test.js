import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import { shallow } from 'enzyme';
import HomeScreen from '../screens/home/home';
import StageSelectScreen from '../screens/stageSelect/stageSelect';
import GameScreen from '../screens/game/game';
import RemoteImage from '../components/photo/remoteImage';
import Keyboard from '../components/keys/Keyboard';

const mockedNavigate = jest.fn();

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

describe('Home Screen', () => {
  it('renders correctly', () => {
    const homeScreen = render(<HomeScreen />);
    expect(homeScreen).toBeTruthy();
  });

  it('navigates to the StageSelect screen when the start button is pressed', () => {
    const { getByTestId } = render(<HomeScreen navigation={{ navigate: mockedNavigate }} />);
    const startButton = getByTestId("start-button");

    fireEvent.press(startButton);

    expect(mockedNavigate).toHaveBeenCalledWith('StageSelect');
  });
});

describe('Stage Select Screen', () => {
  it('renders correctly', () => {
    const stageSelectScreen = render(<StageSelectScreen />);
    expect(stageSelectScreen).toBeTruthy();
  });
});

describe('Game Screen', () => {
  const mockedParams = {
    route: { params: { words: [{ word: "ant" }], stage: 1 } },
    navigation: ''
  };

  it('renders correctly', () => {
    const gameScreen = render(<GameScreen {...mockedParams} />);
    expect(gameScreen).toBeTruthy();
  });

  it('should add 100 points when the answer is correct', () => {
    const { getByTestId } = render(<GameScreen {...mockedParams} />);

    fireEvent.press(getByTestId("a"));
    fireEvent.press(getByTestId("n"));
    fireEvent.press(getByTestId("t"));
    fireEvent.press(getByTestId("↵"));

    expect(getByTestId("points").props.children).toEqual([100, " pts"]);
  });

  it('should deduct 100 points when the answer is incorrect', () => {
    const { getByTestId } = render(<GameScreen {...mockedParams} />);

    fireEvent.press(getByTestId("a"));
    fireEvent.press(getByTestId("n"));
    fireEvent.press(getByTestId("d"));
    fireEvent.press(getByTestId("↵"));

    expect(getByTestId("points").props.children).toEqual([-100, " pts"]);
  });
});

describe('Remote Image', () => {
  it('displays the image correctly', () => {
    const source = { uri: 'https://example.com/image.png' };
    const { getByTestId } = render(<RemoteImage source={source} />);
    const image = getByTestId('photo-frame-image');
    expect(image.props.source).toEqual(source);
  });
});

describe('Keyboard', () => {
  const onKeyPressed = jest.fn();
  const wrapper = shallow(<Keyboard onKeyPressed={onKeyPressed} />);

  it('calls onKeyPressed when a key is pressed', () => {
    const key = 'a';
    const touchable = wrapper.findWhere(node => node.prop('testID') === key);
    touchable.props().onPress();
    expect(onKeyPressed).toHaveBeenCalledWith(key);
  });
});


