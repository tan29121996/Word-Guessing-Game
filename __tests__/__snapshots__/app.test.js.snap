// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keyboard renders correctly 1`] = `ShallowWrapper {}`;
